### DISCUSSION TITLE
How to securily use access tokens?

### DISCUSSION THREAD

**Discussion by buhtz** [PyGithub/PyGithub#2255] [Q&A]
I want to use PyGithub to generate some statistics about specifics project where I am one of the owners/maintainers.

Do I have to authenticate via PyGithub to get the data (e.g. Issues)? If I would use the browser I can "read" the Issue list with my eyes of every public repo without being logged in. So where is the difference?

If I have to authenticate...

It is clear (to me) that I can not hard code a password or any other credential into the python code. So what can I do?

1. I am not able to use uername and password because my GitHub password is autogenerated by keepassxc and 999 chars long. And I won't copy and paste it every time because it is uncomfortable and insecure.

What is an alternative? I don't see a difference between password and access token. The letter is also a string I have to "remember" or store somewhere.

How can I solve this without compromising my security?

**Comment by EnricoMi** [comment]
The username and password should not be used as this has full control over your github account. Compromising these credentials could lead to loss of your account. Better create a token with minimal scope / permissions and use that.

However, any credentials (password or token) could be stored in a file that is permissioned for your user only. Store this file our side your code, so you don't accidentally commit this to a repository.

Then read this in Python as follows:

    with open("credentials.txt") as r:
        token = r.readline().strip()

    auth = github.Auth.Token(token)
    g = github.Github(auth=auth)

**Comment by EdmundGoodman** [comment]
For posterity, another way of doing this is using a `.env` file (which should be excluded in your `.gitignore` file) along with the [python-dotenv](https://pypi.org/project/python-dotenv/) library. This is consistent with the way people tend to do things when storing secrets for Docker and other similar technologies which need access to secrets, such as tokens.

Instantiating the class would then look like this:

```python
from dotenv import dotenv_values
from github import Auth, Github

auth_token = dotenv_values()["GITHUB_API_TOKEN"]
if auth_token is None:
    raise RuntimeError("Unable to authenticate to GitHub!")
github_api = Github(auth=Auth.Token(auth_token))
```

Along with an appropriate `.env` file like this:

```.env
GITHUB_API_TOKEN=<your_secret_token_here>
```
